File: 2024/Mergesort.java
import java.util.Random;

public class Mergesort {
    private int[] data;
    private int[] arr;

    private int number;

    public void sort(int[] values) {
        this.data = values;
        number = values.length;
        this.arr = new int[number];
        mergesort(0, number - 1);
    }

    private void mergesort(int start, int end) {
        // Verifica se start é menor que end, caso contrário terminou
        if (start < end) {
            // Obtém o índice o elemento no meio
            int mid = (start + end) / 2;
            // Ordena o lado esquerdo
            mergesort(start, mid);
            // Ordena o lado direito
            mergesort(mid + 1, end);
            // Combina ambos
            merge(start, mid, end);
        }
    }

    private void merge(int start, int mid, int end) {
        int p = start, q = mid+1;
        int k=0;
        for(int i=start; i<= end; i++) {
            if(p > mid)
                arr[k++] = data[q++];
            else if(q > end)
                arr[k++] = data[p++];
            else if(data[p] < data[q])
                arr[k++] = data[p++];
            else
                arr[k++] = data[q++];
        }
        for(p=0; p<k; p++)
            data[start++] = arr[p];
    }

    public static void main(String args[]) {
        Random r = new Random();
        int MAX = 10000;
        int[] data = new int[MAX];
        for(int i=0; i<data.length; i++)
            data[i] = r.nextInt(data.length*10);

        long start = System.nanoTime();
        Mergesort ms = new Mergesort();
        ms.sort(data);
        long end = System.nanoTime();
        for(int i=0; i<data.length; i++)
            System.out.print(data[i]+" ");
        System.out.println();
        System.out.printf("Tempo para ordenar %d elementos: %d ns\n",MAX,end-start);
    }
}

File: 2024/March/C/mergesort.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define MAX 10000

int arr[MAX];

void sort(int data[], int total);
void mergesort2(int data[], int start, int end);
void merge(int data[], int start, int mid, int end);

void sort(int data[], int total) {
    mergesort2(data, 0, total-1);
}

void mergesort2(int data[], int start, int end) {
    if (start < end) {
        int mid = (start + end) / 2;
        mergesort2(data, start, mid);
        mergesort2(data, mid + 1, end);
        merge(data, start, mid, end);
    }
}

void merge(int data[], int start, int mid, int end) {
    int p = start, q = mid+1;
    int k=0;
    for(int i=start; i<= end; i++) {
        if(p > mid)
            arr[k++] = data[q++];
        else if(q > end)
            arr[k++] = data[p++];
        else if(data[p] < data[q])
            arr[k++] = data[p++];
        else
            arr[k++] = data[q++];
    }
    for(p=0; p<k; p++)
        data[start++] = arr[p];
}

int main()
{
    int data[MAX];
    for(int i=0; i<MAX; i++)
        data[i] = rand()%(MAX*10);
    long start = clock();
    sort(data, MAX);
    long end = clock();
    //for(int i=0; i<MAX; i++)
    //    printf("%d ", data[i]);
    //printf("\n");
    printf("Tempo para %d elementos: %ld ns\n", MAX,(end-start));
}
